Step3 리뷰 요청합니다!

jisu1211 opened the pull request:
안녕하세요~ 
2단계에서 세심한 피드백 감사했습니다 ㅎㅎ
피드백 주신 부분들 수정하였고, 
3단계도 열심히 구현해봤는데 피드백 부탁드립니다!
감사합니다 😊

jisu1211 committed:
refactor:  2단계 피드백 반영
- 테스트 클래스/메소드에서 public 접근제어자 제거
- 분기 단위별 줄바꿈
- convertStrToInt -> convertToPositiveInt 메서드 이름 구체화
- @DisplayName으로 테스트 의도 구체화
- 요구사항 "숫자 하나를 문자열로 입력할 경우 해당 숫자를 반환한다."가 제대로 구현되지 않은 부분 수정

jisu1211 committed:
docs: 로또(자동) README 작성
refactor:
- RuntimeException -> IllegalArgumentException으로 구체화

jisu1211 committed:
feat: view 구현
- Console
   - LottoInput/LottoOutput을 이용해 사용자로부터 값을 입력받고, 결과를 출력할 수 있다.
   - LottoInput: 사용자의 입력받기를 담당하는 클래스
   - LottoOutput: 결과의 출력을 담당하는 클래스
   - 사용자가 원하는 개수만큼 로또를 구입할 수 있다.
- Order: 사용자의 로또 주문 정보를 관리하는 클래스(구매 수량/총 금액)
  - 구매 수량이 음수 혹은 0이면 IllegalArgumentException이 발생한다.

test: Order 테스트 코드 추가

jisu1211 committed:
feat: Store 기능 구현
- Store
  - 로또 숫자의 범위(1-45)에 해당하는 랜덤한 숫자 6개를 사용자가 원하는 개수만큼 구입할 수 있다.
  - 현재 로또의 가격을 가진다.

jisu1211 committed:
feat: Lottos, Lotto 기능 구현
- Lottos
  - 사용자가 구매한 로또의 목록을 가진다.
  - 로또의 추첨을 요청한다(Lottos -> WinningNumbers).
- Lotto
  - 로또 1개를 의미하며, 1-45 범위 안의 랜덤한 숫자 6개를 가진다.
  - 지난 주 당첨 번호를 입력받아 당첨 상태(1-4등/낙첨)를 가진다.

test: Lotto 테스트 추가

jisu1211 committed:
feat: WinningNumbers 구현
- WinningNumbers
  - 지난 주 당첨 번호를 입력받을 수 있다.
  - 지난 주 당첨 번호에 음수 혹은 0이 포함되어 있으면 IllegalArgumentException이 발생한다.
  - 지난 주 당첨 번호의 개수가 6개가 아니라면 IllegalArgumentException이 발생한다.
  - 각 Lotto 객체에 지난 주 당첨 번호를 넘겨주며 추첨을 요청한다.

jisu1211 committed:
feat: WinStats 구현
- WinStats: 아래와 같은 클래스들과 연관관계를 맺고 있는 클래스
   - ProfitRate: 누적 당첨 금액/구매 수량을 입력받아 수익률을 계산하는 클래스
   - AmountByLottoStatus: 누적 당첨 금액과 총 당첨 금액, 각 등수별 당첨 개수를 계산하는 클래스

test: ProfitRate/AmountByLottoStatus 테스트 추가

limwoobin commented on README.md:
```diff
@@ -14,11 +14,68 @@
+4개 일치 (50000원)- 0개\
+5개 일치 (1500000원)- 0개\
+6개 일치 (2000000000원)- 0개\
+총 수익률은 0.35입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)
```
요구사항 정리 👍

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Lotto {
+
+    private final List<Integer> numbers = new ArrayList<>();
+
+    public Lotto(List<Integer> numbersFromStore) {
+        numbers.addAll(numbersFromStore);
+    }
```
```java
private final List<Integer> numbers;

public Lotto(List<Integer> numbersFromStore) {
    numbers = .numbersFromStore;
}
```
와 같이 객체 생성시에 생성자에서 바로 할당하는건 어떨까요??
리스트를 선언하고 객체를 add하는것보다 간소화할 수 있을것같아요!

jisu1211 replied:
네~ 제안해주신 부분이 더 간단해보이네요! 수정하도록 하겠습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,35 @@
+    private int matchNumbers(List<Integer> winningNumbers) {
+        int result = NumberUtil.INIT_ZERO;
+        for(Integer num : numbers) {
+            result += Collections.frequency(winningNumbers, num);
```
로또번호가 있는경우 result에 더해주는군요
그런데 frequency의 경우 해당 값이 몇 개가 있는지의 목적으로 사용되는게 더 알맞다고 생각들어
contains를 이용해하는건 어떨까 생각듭니다!

이 부분은 개인적인 의견이니 참고만 해주세요!

jisu1211 replied:
저도 저 부분 작성하면서 애매하다고 생각했는데, 말씀해주신 내용으로 변경해보도록 하겠습니다 ☺️

limwoobin commented on src/main/java/study/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+    private final List<Lotto> lottos = new ArrayList<>();
+    private final WinStats stats;
+
+    public Lottos(List<List<Integer>> allNumbersFromStore) {
+        addLottos(allNumbersFromStore);
+        stats = new WinStats(lottos.size());
+    }
```
생성자로 `List<List<Integer>>` 를 사용하시는것보다는 어느 객체를 인자로 전달받는지 보다 명확한
`List<Lotto>` 는 어떨까요?? 
해당 객체를 사용하는쪽에서 보면 복잡도가 높아보일수도 있을것같아서요!

jisu1211 replied:
`List<Lotto>`로 변경하니까 훨씬 보기에 간단해지네요! 좋은 피드백 감사합니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/Store.java:
```diff
@@ -0,0 +1,41 @@
+    private static final int LOTTO_STAT_IDX = 0;
+    private static final int LOTTO_END_IDX = 6;
+
+    private static final List<Integer> rangeOfLottoNumbers = getRangeOfLottoNumbers();
+
+    private static List<Integer> getRangeOfLottoNumbers() {
+        return IntStream.rangeClosed(LOTTO_MIN_NUM, LOTTO_MAX_NUM)
+                .boxed()
+                .collect(Collectors.toList());
+    }
```
로또번호를 캐싱해서 추출하셧군요 굉장히 좋네요 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private List<Integer> inputWinningNumbers(String winningNumbers) {
+        String[] strings = Splitter.split(winningNumbers);
```
strings보다는 입력받은 당첨번호인것을 더 드러내는 네이밍이 오면 좋겠습니다!

limwoobin commented on src/main/java/study/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,50 @@
+
+    private List<Integer> inputWinningNumbers(String winningNumbers) {
+        String[] strings = Splitter.split(winningNumbers);
+        if(checkSize(strings)) {
+            return convertStringArrToList(strings);
+        }
+        return Collections.EMPTY_LIST;
```
checkSize에서 예외를 바로 던져버리기때문에 
맨 아래의 return Collections.EMPTY_LIST는 실행될 일이 없을것같아요!

checkSize를 void로 변경해 아래와 같이 사용하는건 어떻게생각하시나요??
```java
checkSize(strings);
return convertStringArrToList(strings);
```

jisu1211 replied:
제가 작성한 부분에서 말씀해주신 대로 마지막 return은 의미가 없어보이네요 ㅎㅎ 수정하겠습니다!

limwoobin commented on src/main/java/study/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+    public String toString() {
+        StringBuffer buffer = new StringBuffer();
+        lottos.forEach((lotto) -> {
+            buffer.append(lotto.toString() + "\n");
```
StringBuffer를 쓰셧으니 통일감있게
`buffer.append(lotto.toString()).append("\n");` 이렇게 쓰는건 어떨까요??

limwoobin commented on src/test/java/study/lotto/domain/WinningNumbersTest.java:
```diff
@@ -0,0 +1,36 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class WinningNumbersTest {
```
저의 경우 테스트코드 작성시 제일 우선시하는게 
어느 기능에 대한 테스트인지 명확하게 나타내는것입니다. 
메소드에도 꼼꼼하게 작성해주셧지만 클래스에도 `@DisplayName("당첨번호 테스트")` 와 같은 description을 남겨주는것도 좋은방법입니다!

이 부분은 개인의견이라 참고만 해주세요!

limwoobin replied:
<img width="490" alt="image" src="https://user-images.githubusercontent.com/28802545/198866204-c17557f2-a815-401a-abd8-138b85fc4f17.png">

테스트 수행시 테스트 케이스가 전보다 명확해진것같네요 👍

limwoobin commented on src/main/java/study/lotto/Application.java:
```diff
@@ -0,0 +1,10 @@
+import study.lotto.view.Console;
+
+public class Application {
+    public static void main(String[] args) {
+        Console console = new Console();
+        console.startConsole();
+    }
+}
```
로또게임의 시작인에 단순 Console이라기엔 명확하게 와닿지가 않네요
`LottoController or LottoGames` 와 같은 명시적인 이름은 어떨까요?

limwoobin commented on src/main/java/study/lotto/domain/WinStats.java:
```diff
@@ -0,0 +1,32 @@
+        profitRate.calculate(amountByLottoStatus.sumTotalAmount());
+    }
+
+    @Override
+    public String toString() {
+        StringBuffer sb = new StringBuffer("\n당첨 통계");
+        sb.append("\n---------");
+        sb.append("\n3개 일치 (5000원) - " + amountByLottoStatus.countByLottoStatus(LottoStatus.FOURTH_PLACE));
+        sb.append("\n4개 일치 (50000원) - " + amountByLottoStatus.countByLottoStatus(LottoStatus.THIRD_PLACE));
+        sb.append("\n5개 일치 (1500000원) - " + amountByLottoStatus.countByLottoStatus(LottoStatus.SECOND_PLACE));
+        sb.append("\n6개 일치 (2000000000원) - " + amountByLottoStatus.countByLottoStatus(LottoStatus.FIRST_PLACE));
+        sb.append("\n총 수익률은 " + profitRate.toString() + "입니다.");
+        return sb.toString();
+    }
```
해당 문구는 view영역에서 하는개 알맞지 않을까요??

limwoobin commented on src/test/java/study/lotto/domain/ProfitRateTest.java:
```diff
@@ -0,0 +1,16 @@
+
+class ProfitRateTest {
+
+    @Test
+    void 수익률_계산() {
+        ProfitRate profitRate = new ProfitRate(14);
+        profitRate.calculate(5000);
+
+        assertEquals("0.36", profitRate.toString());
+    }
```
`@ParameterizedTest` 을 이용해 다양한 수익률을 계산해보는것도 좋아보입니다!

limwoobin reviewed (changes requested):
안녕하세요 지수님 
2단계때 말씀드린 내용 피드백 잘 해주셨네요 ㅎㅎ
3단계 미션 또한 굉장히 잘 수행해주셨습니다! 객체를 나누는 부분이 굉장히 인상적이었습니다. 

다만 같이 고민해보면 좋을만한 내용이 있어 코멘트 드렸습니다
확인 후 다시 요청부탁드립니다!

+ 추가로 테스트코드에 대해서는 도메인의 클래스 단위로 작성해보면서 연습하는것을 추천드립니다! 이번 경우엔 아쉽게도 몇몇클래스가 빠져있었네요ㅠ 이부분도 같이 확인부탁드려요 ㅎㅎ

jisu1211 committed:
refactor: 3단계 피드백 반영
- 생성자에서 List 사용시 addAll 사용하는 부분을 바로 변수에 대입하는 방식으로 변경
- Collections.frequency -> ArrayList의 contains로 변경
- List<List<Integer>> -> List<Lotto>로 간략하게 변경
- 의미있는 변수 이름으로 변경
- 로직 상 실행될 일 없는 코드 제거

test: 검증하지 않은 테스트 코드 추가, 테스트 클래스에 @DisplayName을 추가

jisu1211 commented:
피드백 주신 부분들 수정했는데 다시 리뷰 부탁드리겠습니다!
반영하고나니 코드가 훨씬 깔끔해졌네요 ㅎㅎ 감사합니다👍

jisu1211 committed:
fix: 요구사항에서 부족했던 사항 구현
- 사용자가 구매한 로또 목록이 출력될 때 정렬되지 않았던 부분 수정
- 사용자가 기본 로또 가격 이하로 금액을 입력했을 때 IllegalArgumentException 발생하도록 수정
- 사용자가 입력한 지난주 당첨 번호에 중복된 숫자가 입력되지 못하도록 수정

feat: 새로운 객체 추가
- Lotto에 입력되는 숫자들에 대한 유효성 체크를 용이하게 하기위한 LottoNumber 클래스 추가
- 지난 주 당첨 숫자로 Lotto와 같은 형식으로 관리할 수 있도록 WinningLotto 클래스 추가

test: 코드 수정에 따라 테스트 변경, WinningLottoTest, LottoNumberTest 추가

jisu1211 commented:
코드를 다시 살펴보니, 요구사항에서 부족했던 부분들이 보여서 코드를 수정했습니다!
그리고 `로또 피드백` 영상을 미리 시청하고, 좀 더 개선하고 싶은 부분들 반영했습니다 ㅎㅎ
아직 많이 익숙하지 않아 어렵네요 ㅠㅠ 감사합니다 ☺️

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,55 @@
+        this(toSet(numbersFromStore));
+    }
+
+    private static Set<LottoNumber> toSet(List<Integer> numbersFromStore) {
```
static으로 사용한 이유가 따로 있을까요??~

jisu1211 replied:
다른 생성자를 사용하려고 static을 사용했는데 생각해보니 static을 사용하지 않고도 해결이 가능했네요!
`this.numbers = toSet(numbersFromStore);`
이렇게 수정해보겠습니다!

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static Set<LottoNumber> toSet(List<Integer> numbersFromStore) {
+        return numbersFromStore.stream()
+                .map((num) -> LottoNumber.of(num))
```
메서드참조를 이용해 `.map(LottoNumber::of)` 로 사용할 수 있겠군요

limwoobin commented on src/main/java/study/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,54 @@
+        this.lottoNumber = lottoNumber;
+    }
+
+    public static LottoNumber of(int number) {
+        if(Objects.isNull(lottoNumbers.get(number))) {
+            throw new IllegalArgumentException("[ERROR] the given number is an invaild lotto number.");
+        }
+        return lottoNumbers.get(number);
+    }
```
코드 전체적으로 자바 코드컨벤션을 한번 확인해주세요! (줄개행, space 등등)

jisu1211 replied:
그 부분을 자세히 보지 못했네요,, 부족한 점 짚어주셔서 감사합니다~

limwoobin commented on src/main/java/study/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,26 @@
+    private final List<Lotto> lottos;
+    private final WinStats stats;
+
+    public Lottos(List<Lotto> allNumbersFromStore) {
+        lottos = allNumbersFromStore;
+        stats = new WinStats(lottos.size());
+    }
```
해당 클래스의 인스턴스변수는 this 를 붙여주시는걸 추천드려요!
그래야 해당 클래스에서 쓰이는 변수라는걸 명확히 할 수 있습니다

limwoobin commented on src/main/java/study/lotto/view/LottoGames.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Map;
+
+public class LottoGames {
```
LottoGames는 어플리케이션에서 게임실행을 담당하는 핵심 클래스인데
view영역에 있는게 맞을까 하는 의문이 있네요 ㅎㅎ 지수님은 어떻게 생각하시나요?

jisu1211 replied:
아직 클래스들을 패키지로 분류할 때 어떤 패키지가 적정할지 고민이 많이 됩니다ㅠㅠ 
우빈님은 어디가 적정하다고 생각하시는지 도움 주실 수 있으실까요..? 😢

limwoobin reviewed (approved):
안녕하세요 지수님 피드백드린 내용 꼼꼼히 잘 반영해주셨네요
테스트 코드도 굉장히 촘촘합니다! ㅎㅎ
요구사항을 충족하지 못한 부분 또한 잘 반영해주셨습니다

다음미션 시작할때 같이 확인하면 좋을 코멘트 몇개 드렸습니다
이번미션은 여기서 merge 하겠습니다 
다음미션도 화이팅입니다!

+ 궁금하신 내용이나 논의하고픈 내용이 있으면 편하게 코멘트나 dm으로 문의주세요!
